src/Microsoft.Identity.Client/Instance/AuthorityEndpointResolutionManager.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.AppConfig;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Exceptions;
+using Microsoft.Identity.Client.OAuth2;
+
+namespace Microsoft.Identity.Client.Instance
+{
+    internal class AuthorityEndpointResolutionManager : IAuthorityEndpointResolutionManager
+    {
+        private static readonly ConcurrentDictionary<string, AuthorityEndpointCacheEntry> EndpointCacheEntries =
+            new ConcurrentDictionary<string, AuthorityEndpointCacheEntry>();
+
+        private readonly IServiceBundle _serviceBundle;
+
+        public AuthorityEndpointResolutionManager(IServiceBundle serviceBundle, bool shouldClearCache = true)
+        {
+            _serviceBundle = serviceBundle;
+            if (shouldClearCache)
+            {
+                EndpointCacheEntries.Clear();
+            }
+        }
+
+        public async Task<AuthorityEndpoints> ResolveEndpointsAsync(
+            AuthorityInfo authorityInfo,
+            string userPrincipalName,
+            RequestContext requestContext)
+        {
+            if (authorityInfo.AuthorityType == AuthorityType.Adfs && string.IsNullOrEmpty(userPrincipalName))
+            {
+                throw MsalExceptionFactory.GetClientException(
+                    CoreErrorCodes.UpnRequired,
+                    CoreErrorMessages.UpnRequiredForAuthroityValidation);
+            }
+
+            if (TryGetCacheValue(authorityInfo, userPrincipalName, out AuthorityEndpoints endpoints))
+            {
+                requestContext.Logger.Info("Resolving authority endpoints... Already resolved? - TRUE");
+                return endpoints;
+            }
+
+            requestContext.Logger.Info("Resolving authority endpoints... Already resolved? - FALSE");
+
+            var authorityUri = new Uri(authorityInfo.CanonicalAuthority);
+            string path = authorityUri.AbsolutePath.Substring(1);
+            string tenant = path.Substring(0, path.IndexOf("/", StringComparison.Ordinal));
+            bool isTenantless = Authority.TenantlessTenantNames.Contains(tenant.ToLowerInvariant());
+
+            // TODO: where is the value in this log message?  we have a bunch of code supporting printing just this out...
+            requestContext.Logger.Info("Is Authority tenantless? - " + isTenantless);
+
```

Comment:
yeah...i don't find this log message helpful when helping customers. maybe you could reduce in footprint.

Reply:
i dislike this comment as well and the whole concept of tenantless.  will fix as part of todo cleanup.